src/journal/src/Journal/Internal/Metrics.hs

```diff
@@ -16,85 +20,171 @@ import Journal.Internal.ByteBuffer
 
 ------------------------------------------------------------------------
 
-data Metrics a = Metrics
-  { mMetadata :: ByteBuffer
-  , mBuffer   :: ByteBuffer
-  }
 
-newtype MetricsSchema a = MetricsSchema
-  { unMetricsSchema :: [(a, MetricsType)]
+data Metrics c h = Metrics
+  { mCounterBuffer   :: ByteBuffer
+  , mHistogramBuffer :: ByteBuffer
   }
 
-data MetricsType = Counter | Histogram
+data MetricsSchema c h = MetricsSchema
+  { msVersion :: Int
+  }
 
-newMetrics :: (Enum a, Bounded a) => MetricsSchema a -> FilePath -> IO (Metrics a)
-newMetrics ms@(MetricsSchema xs) fp = do
-  bb <- mmapped fp (sizeOfOffsets + sizeOfMetrics ms)
-  meta <- wrapPart bb 0 sizeOfOffsets
-  buf  <- wrapPart bb (sizeOfOffsets + 1) (sizeOfOffsets + sizeOfMetrics ms)
-  return (Metrics meta buf)
+-- TODO have a header in the file and check that schema is the same as existing in file?
+newMetrics :: forall c h. (Enum c, Bounded c, Enum h, Bounded h) => MetricsSchema c h -> FilePath -> IO (Metrics c h)
+newMetrics _ fp = do
+  -- bb <- mmapped fp (sizeOfCounters + sizeOfHistograms) -- mmap seems problematic
+  bb <- allocate (sizeOfCounters + sizeOfHistograms)
+  cbuf <- wrapPart bb 0 sizeOfCounters
+  hbuf <- wrapPart bb sizeOfCounters sizeOfHistograms
+  return (Metrics cbuf hbuf)
   where
-    sizeOfOffsets = length xs * sizeOf (8 :: Int)
-
--- XXX: how can we avoid `incrCounter` being called on a histogram?
-incrCounter :: (Enum a, Bounded a) => Metrics a -> a -> Int -> IO ()
-incrCounter = undefined
-
-measure :: (Enum a, Bounded a) => Metrics a -> a -> Int -> IO ()
-measure = undefined
+    allCounterLabels :: [c]
+    allCounterLabels = enumFrom minBound
+    allHistogramLabels :: [h]
+    allHistogramLabels = enumFrom minBound
+    sizeOfCounters = length allCounterLabels * sizeOfACounter
```

Comment:
We can avoid building the list using: `sizeOfCounters = (fromEnum (maxBound :: c) + 1) * sizeOfCounter`

Reply:
Thanks